src/liballoc/rc.rs

```diff
@@ -548,13 +721,44 @@ impl<T: ?Sized> Rc<T> {
     pub fn get_mut(this: &mut Self) -> Option<&mut T> {
         if Rc::is_unique(this) {
             unsafe {
-                Some(&mut this.ptr.as_mut().value)
+                Some(Rc::get_mut_unchecked(this))
             }
         } else {
             None
         }
     }
 
+    /// Returns a mutable reference to the inner value,
+    /// without any check.
+    ///
+    /// See also [`get_mut`], which is safe and does appropriate checks.
+    ///
+    /// [`get_mut`]: struct.Rc.html#method.get_mut
+    ///
+    /// # Safety
+    ///
+    /// There must be no other `Rc` or [`Weak`] pointers to the same value.
+    /// This is the case for example immediately after `Rc::new`.
```

Comment:
Crucially, also, not such pointers must be created until the lifetime of the returned reference ends!

Reply:
Good point, but since the returned reference borrows `&mut Rc<T>` for the only `Rc` pointing to this `T`, isn’t it impossible to create such pointers?

Reply:
Oh I see you demand this being the only `Rc` at all. Why that? If the others are all "not used", there shouldn't be a problem...

Reply:
Oh indeed maybe. But what does "not used" mean exactly, for a call to this method to be valid?

Reply:
Not deref'd. So, there must be no other reference pointing to the inner value (such as one obtains via `deref`, `get_mut`, `get_mut_unchecked`, `make_mut`, maybe more).

Reply:
Done.

Reply:
Just to note: this exposes shared internal mutability to (`A`)`Rc` where the stable API only exposes (runtime checked) unshared internal mutability (of the payload).

Reply:
@CAD97 I am not entirely sure what you mean... I do not see any interior mutability here. That term is specifically tied to `UnsafeCell` in Rust, and there is none here.

Would you say that the fact that we can create a raw pointer into a `Box` "exposes shared internal mutability"? Because this operation is very similar.

Reply:
I think another way to phrase @CAD97’s comment/concern is: given `let some_box: Box<u32> = /*…*/; let ptr = &mut *some_box;`, under what conditions if any is it safe to write through `ptr` while `some_box` still exists (and therefore "shares" the same `u32` value by having another pointer to it).

Compare with `Arc::get_mut` which ensures there’s only one reference to the value and therefore none of this kind of "sharing".

Reply:
> under what conditions if any is it safe to write through ptr while some_box still exists (and therefore "shares" the same u32 value by having another pointer to it).

That goes deep into the aliasing model -- and is not specific to `Box` or `Arc` at all, I would say.

Your example uses a safe reference for `ptr` so you won't be able to do anything illegal, but assuming you meant `&mut *some_box as *mut _`, then the answer that Stacked Borrows gives is that you may use `ptr` until you do *anything* with `some_box` the next time -- including things like moving it or borrowing from it.

Reply:
> not specific to `Box` or `Arc` at all

Yeah a code sample just a way to express something without terminology

> assuming you meant `&mut *some_box as *mut _`

Oops, yes I did.

----

Ok I understand you might not have a final answer yet for some cases of aliasing rules. So the question is: does documenting `Arc::get_mut_unchecked` as safe to use while other `Arc` references to the same value exist (just not deref’ed) constrain a future decision we might want to make for aliasing rules? As opposed to documenting that it may only be used if only one `Arc` and no `Weak` exist.

Reply:
Oh, now I see what you are getting at. Good question.

This can only become a problem if `Arc`/`Rc` is ever understood by the optimizer to mean something. With `Box`, it is basically a primitive type and also uses `Unique`, hence the strict rules; for `Arc` that seems unlikely.

> while other Arc references to the same value exist (just not deref’ed)

Deref would certainly be bad, but even cloning creates a reference to the `ArcInner`, right? None of these should be allowed.

I think we'd be on the safe side if we said that other `Arc` might exist but nothing may be done with them (not even a move), similar to `Box`.